File: 839. Similar String Groups.cpp
// Union Find
// Time: O(n^2 * l), space: O(n)
// but still extremely slow
class Solution {
private:
    map<int, int> parent;
    int group;
    
    // w1 and w2 are already anagrams
    bool similar(string &w1, string &w2) {
        int diffs = 0;
        for (int i = 0; i < w1.size(); ++i) {
            if (w1[i] != w2[i]) diffs += 1;
            if (diffs > 2) return false;
        }
        
        return true;
    }
    
    int find(int x) {
        if (parent.count(x) == 0) parent[x] = x;
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }
    
    void unite(int x, int y) {
        int rootx = find(x), rooty = find(y);
        if (rootx != rooty) {
            group -= 1;
            parent[rootx] = rooty;
        }
    }
    
public:
    int numSimilarGroups(vector<string>& A) {
        int n = A.size();
        parent = map<int, int>();
        group = n;
        
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                if (similar(A[i], A[j])) unite(i, j);
            }
        }
        
        return group;
    }
};

File: 839. Similar String Groups.py
# Union Find
# Time: O(n^2 * l), space: O(n)
# though this implementation in python cannot AC on leetcode, always TLE
# but exactly the same strategy passes in C++
class Solution:
    def numSimilarGroups(self, A: List[str]) -> int:

        # w1 and w2 are already anagrams
        def similar(w1, w2):
            if (w1 == w2): return True
            if (len(w1) != len(w2)): return False
            diffs = []
            for i in range(len(w1)):
                if (w1[i] != w2[i]): diffs.append(i)
                if (len(diffs) > 2): return False
            i, j = diffs
            return [w1[i], w1[j]] == [w2[j], w2[i]]

        parent = {}
        self.group = len(A)

        def find(x):
            if (x not in parent): parent[x] = x
            if (parent[x] != x):
                parent[x] = find(parent[x])
            return parent[x]

        def union(x, y):
            rootx, rooty = find(x), find(y)
            if (rootx != rooty):
                self.group -= 1
                parent[rootx] = rooty

        for i, w in enumerate(A):
            for j in range(i + 1, len(A)):
                if (similar(w, A[j])):
                    union(i, j)
        return self.group
